### DISCUSSION TITLE
Error occuring when executing Payrun job

### DISCUSSION THREAD

**Discussion by Abdullahifisol** [Payroll-Engine/PayrollEngine#4] [Q&A] [answered]
Hi, @Giannoudis , Hope you'r doing well

So, the error is coming when I generate a build of backend APIs and deploy it on our server and execute the whole process step by step but error shows up when executing payrun job.

The strange thing is that this error doesn't show up when I execute through your beta release build (PayrollEngineSetup_0.8.0-beta.1). I think it's missing something from this Backend Repository.

I tried several times but still the same but not with your beta build.
This is happening on Local as well.
![image](https://github.com/Payroll-Engine/PayrollEngine/assets/96053562/991beafd-ecca-4691-9f4a-a797b48e3cf5)

**Comment by Giannoudis** [comment]
Can you give me a JSON example?

**Comment by Abdullahifisol** [comment]
{
  "$schema": "../../Schemas/PayrollEngine.Exchange.schema.json",
  "createdObjectDate": "2023-01-01T00:00:00.0Z",
  "tenants": [
    {
      "identifier": "WeekSimpleTenant",
      "culture": "en-US",
      "users": [
        {
          "identifier": "lucy.smith@foo.com",
          "firstName": "Lucy",
          "lastName": "Smith",
          "culture": "en-US",
          "userType": "Supervisor"
        }
      ],
      "calendars": [
        {
          "name": "WeekCalendar",
          "cycleTimeUnit": "Year",
          "periodTimeUnit": "Week"
        }
      ],
      "calendar": "WeekCalendar",
      "divisions": [
        {
          "name": "WeekSimpleDivision"
        }
      ],
      "employees": [
        {
          "identifier": "višnja.müller@foo.com",
          "firstName": "Višnja",
          "lastName": "Müller",
          "divisions": [
            "WeekSimpleDivision"
          ]
        }
      ],
      "regulations": [
        {
          "name": "WeekSimplePayroll",
          "description": "The base of all payrolls",
          "cases": [
            {
              "name": "AHV",
              "caseType": "National",
              "fields": [
                {
                  "name": "AHV AN",
                  "valueType": "Percent",
                  "timeType": "Period"
                }
              ]
            },
            {
              "name": "AccidentInsurance",
              "caseType": "Company",
              "fields": [
                {
                  "name": "NBU Set",
                  "valueType": "Percent",
                  "timeType": "Period"
                }
              ]
            },
            {
              "name": "WeeklyWage",
              "caseType": "Employee",
              "fields": [
                {
                  "name": "WeeklyWage",
                  "valueType": "Money",
                  "timeType": "CalendarPeriod"
                }
              ]
            },
            {
              "name": "EmploymentLevel",
              "caseType": "Employee",
              "fields": [
                {
                  "name": "EmploymentLevel",
                  "valueType": "Percent",
                  "timeType": "Period"
                }
              ]
            },
            {
              "name": "Department",
              "caseType": "Employee",
              "fields": [
                {
                  "name": "Department",
                  "valueType": "String",
                  "timeType": "Period"
                }
              ]
            },
            {
              "name": "Bonus",
              "caseType": "Employee",
              "fields": [
                {
                  "name": "Bonus",
                  "valueType": "Money",
                  "timeType": "Moment"
                }
              ]
            }
          ],
          "collectors": [
            {
              "name": "AHV Base"
            }
          ],
          "wageTypes": [
            {
              "wageTypeNumber": 101,
              "name": "WeeklyWage",
              "valueExpression": "var values = GetCaseValues(\"WeeklyWage\", \"EmploymentLevel\"); return values[\"WeeklyWage\"] * values[\"EmploymentLevel\"]",
              "collectors": [
                "AHV Base"
              ]
            },
            {
              "wageTypeNumber": 103,
              "name": "Bonus",
              "valueExpression": "CaseValue[\"Bonus\"]",
              "collectors": [
                "AHV Base"
              ]
            },
            {
              "wageTypeNumber": 104,
              "name": "AHV AN Abzug",
              "valueExpression": "Collector[\"AHV Base\"] * CaseValue[\"AHV AN\"]"
            },
            {
              "wageTypeNumber": 105,
              "name": "NBU Abzug",
              "valueExpression": "Collector[\"AHV Base\"] * CaseValue[\"NBU Set\"]"
            }
          ]
        }
      ],
      "payrolls": [
        {
          "name": "WeekSimplePayroll",
          "divisionName": "WeekSimpleDivision",
          "layers": [
            {
              "level": 1,
              "priority": 1,
              "regulationName": "WeekSimplePayroll"
            }
          ]
        }
      ],
      "payruns": [
        {
          "payrollName": "WeekSimplePayroll",
          "name": "WeekSimplePayrollPayrun1"
        }
      ]
    }
  ]
}

This is the JSON I'm executing through your command console, i got this JSON from your beta release build. (FYI, it worked fine when executed through your beta release build) 

JSON Response:
![image](https://github.com/Payroll-Engine/PayrollEngine/assets/96053562/cfb8bf19-5068-41af-8e81-02f30edb0590)

**Comment by Giannoudis** [comment]
Hello @Abdullahifisol.

Unfortunately, I cannot reproduce the problem with the attached JSON example. Did you create a special version of the backend server?

**Comment by Abdullahifisol** [comment]
Hi @Giannoudis 

I haven't created any special version. Just cloned and ran the project from this URL  "https://github.com/Payroll-Engine/PayrollEngine.Backend.git".

**Comment by Giannoudis** [comment]
It is not necessary to clone the backend repository to use the Payroll Engine.

To create your own payroll definition, follow these steps
1. Install the Payroll Engine as described in the wiki on the [Setup](https://github.com/Payroll-Engine/PayrollEngine/wiki/Setup) page.
    - NET and SQL Server must be pre-installed.
    - Install the latest version using the `Setup.cmd` command.
2. Start the backend application server with the `BackendServer.cmd` command.
3. Write the payroll definition to a JSON file `Payroll.json`.
4. Create the import command file `Import.cmd`.
    > Run "%PayrollConsole%" PayrollImport Payroll.json
    - The `%PayrollConsole%` environment variable is set after installation.

See `Examples/WeekSimplePayroll/Import.cmd` in your local installation for an example.

In addition to the payroll definition, values (`Examples/WeekSimplePayroll/Import.Values.cmd`) and payroll jobs (`Examples/WeekSimplePayroll/Import.Jobs.cmd`) can be imported in the same way.

Further commands in the WeekSimplePayroll example:
- `Setup.cmd`: Complete payroll setup
- `Delete.cmd`: Delete a payroll
- `Report.cmd`: Create a payroll report

> The [Payroll Console Repository](https://github.com/Payroll-Engine/PayrollEngine.PayrollConsole) describes all the commands that can be run from a command file.

**Comment by Abdullahifisol** [comment]
@Giannoudis Thank you so much for taking the time to respond to all my queries.

For now, yes, I'm using this approach to create and test the payroll engine. However, what should I do if there's a requirement to make some changes in the backend server?

**Reply by Giannoudis** [reply, marked as answer]
The following steps are required to make changes to the backend:
1. Clone the backend [repository](https://github.com/Payroll-Engine/PayrollEngine.Backend).
2. Open the `PayrollEngine.Backend.sln` solution using Visual Studio.
3. Implement the changes in the code.
4. Compile the solution in release mode.
5. Copy/overwrite the binaries files from the `bin\Release` folder to the `PayrollEngine.backend` installation directory.
6. Start the backend server application and the payroll console as described above.

Since the backend server is very generic, I would be interested to know what customizations you would like to make in the backend.
